### PR TITLE
fix(resharding): prevent archival data loss

### PR DESCRIPTION
**Pull request by Wiezzel**
Resharding creates intermediate trie nodes via `retain_split_shard` but never persisted them to `DBCol::TrieChanges`. The cold store copy loop reads that column to discover state nodes to copy, so it missed the resharding nodes entirely. After `gc_parent_shard_after_resharding` range-deleted the parent prefix from hot store, these nodes were lost.

Fix: persist insertions-only TrieChanges for each child shard at the resharding block, and have the cold store copy loop read them at the resharding boundary (first block of new epoch looks back at the previous block's hash).

### CODE REVIEW

File: chain/chain/src/resharding/manager.rs [hunk 1]
```diff
@@ -246,6 +251,18 @@ impl ReshardingManager {
             tries.apply_insertions(&trie_changes, *parent_shard_uid, &mut store_update);
             tries.apply_memtrie_changes(&trie_changes, *parent_shard_uid, block_height);
 
+            // Persist TrieChanges so that cold store can copy the resharding
+            // insertions. We store only insertions (no deletions) because
+            // deletions are not applied during resharding and would corrupt
+            // GC refcounts.
+            if save_trie_changes {
```
Comment (shreyan-gupta) [inline chain/chain/src/resharding/manager.rs]: Why do we have the `save_trie_changes` boolean here and why are we not setting the trie changes every time we do resharding?

File: core/store/src/archive/cold_storage.rs [hunk 2]
```diff
@@ -206,22 +212,44 @@ fn copy_state_from_store(
...
+        }
 
-            total_size += value.len();
-            tracing::trace!(target: "cold_store", pretty_key=?near_fmt::StorageKey(&key), "copying state node to colddb");
-            rc_aware_set(&mut transaction, DBCol::State, key, value);
+        // At resharding boundaries, also copy TrieChanges from the resharding
+        // block (previous block). The resharding creates child shard state by
+        // splitting the parent trie, and those TrieChanges are keyed by
+        // (resharding_block_hash, child_shard_uid).
+        if let Some(resharding_hash_key) = resharding_block_hash {
```
Comment (shreyan-gupta) [inline core/store/src/archive/cold_storage.rs]: I don't like the fact that we have a special case for resharding_block here and we are duplicating the code from above. Why do we have this? Can we avoid it and try to keep the cold store loop with minimal changes?

File: chain/chain/src/resharding/manager.rs [hunk 3]
```diff
@@ -246,6 +251,18 @@ impl ReshardingManager {
...
+                store_update.set_trie_changes(
+                    *new_shard_uid,
+                    block_hash,
+                    &trie_changes.insertions_only(),
```
Comment (shreyan-gupta) [inline chain/chain/src/resharding/manager.rs]: Why are we storing insertions only and not all the changes? I don't think they are explicitly used anywhere downstream right?

Comment (shreyan-gupta) [reply chain/chain/src/resharding/manager.rs]: Again, not a fan of the special behavior without an obvious reason.

File: test-loop-tests/src/tests/resharding_cold_storage.rs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+/// copied there, historical queries eventually fail.
+///
+/// This test:
+/// 1. Runs resharding and captures the child shard TrieChanges (before GC).
+/// 2. Waits for the cold store loop to process past the resharding boundary.
+/// 3. Verifies the cold store's State column contains all resharding trie nodes.
+#[test]
+// TODO(spice-test): Assess if this test is relevant for spice and if yes fix it.
+#[cfg_attr(feature = "protocol_feature_spice", ignore)]
+fn test_resharding_trie_nodes_copied_to_cold_store() {
```
Comment (shreyan-gupta) [inline test-loop-tests/src/tests/resharding_cold_storage.rs:41]: Do we have the negation of the test working? I.e. without our change, is this test failing on missing trie key?

File: test-loop-tests/src/tests/resharding_cold_storage.rs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        for op in trie_changes.insertions() {
+            all_insertions.push((*child_shard_uid, *op.hash()));
+        }
+    }
+    assert!(!all_insertions.is_empty());
+
+    // --- 5. Wait for cold store loop to process past the resharding boundary ---
+    let post_resharding_height = head.height;
+    let target_height = post_resharding_height + EPOCH_LENGTH * (GC_NUM_EPOCHS_TO_KEEP + 2);
+    env.test_loop.run_until(
```
Comment (shreyan-gupta) [inline test-loop-tests/src/tests/resharding_cold_storage.rs]: nit: there's a function to run until height.

File: test-loop-tests/src/tests/resharding_cold_storage.rs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    let resharding_block_hash = loop {
+        let header = client.chain.get_block_header(&block_hash).unwrap();
+        let shard_layout = epoch_manager.get_shard_layout(header.epoch_id()).unwrap();
+        if shard_layout == base_shard_layout {
+            break *header.hash();
+        }
+        block_hash = *header.prev_hash();
+    };
+
+    let old_shard_uids: std::collections::HashSet<ShardUId> =
```
Comment (shreyan-gupta) [inline test-loop-tests/src/tests/resharding_cold_storage.rs]: nit: fully qualified name

**Review by shreyan-gupta — COMMENTED** [review]
Couple of comments

File: test-loop-tests/src/tests/resharding_cold_storage.rs (hunk 4 above)
Comment (Wiezzel) [reply test-loop-tests/src/tests/resharding_cold_storage.rs:41]: Yes.

File: chain/chain/src/resharding/manager.rs (hunk 3 above)
Comment (Wiezzel) [reply chain/chain/src/resharding/manager.rs]: Above we call `trie.apply_insertions()` not `trie.apply_all()`, so I wanted to follow this. Also, from what I understand, the changes we save here are later picked up by the GC, which calls `apply_deletions` – that might lead to unwanted deletions and possibly messing up refcounts (as the comment explains).

File: chain/chain/src/resharding/manager.rs (hunk 1 above)
Comment (Wiezzel) [reply chain/chain/src/resharding/manager.rs]: My mistake, I thought we need to check it here, but it's actually handled internally by `ChainStoreUpdate`. Will remove.

File: core/store/src/archive/cold_storage.rs (hunk 2 above)
Comment (Wiezzel) [reply core/store/src/archive/cold_storage.rs]: Let me check that...

Comment (Wiezzel) [reply core/store/src/archive/cold_storage.rs]: I've merged the two blocks to avoid duplicating logic. See how you like the new version.

File: chain/chain/src/resharding/manager.rs (hunk 1 above)
Comment (Wiezzel) [reply chain/chain/src/resharding/manager.rs]: Removed.

**Review by shreyan-gupta — APPROVED** [review]

File: chain/chain/src/garbage_collection.rs [hunk 7]
```diff
@@ -893,17 +893,36 @@ impl<'a> ChainStoreUpdate<'a> {
...
+                .get_shard_layout(block_header.epoch_id())
+                .expect("epoch info must exist");
+            shard_layout.shard_uids().collect()
+        };
+
         for shard_uid in shard_uids_to_gc {
             let trie_changes_key = get_block_shard_uid(&block_hash, &shard_uid);
-            let trie_changes = self.store().get_ser(DBCol::TrieChanges, &trie_changes_key);
+            let trie_changes: Option<TrieChanges> =
+                self.store().get_ser(DBCol::TrieChanges, &trie_changes_key);
```
Comment (staffik) [inline chain/chain/src/garbage_collection.rs:913]: why this change?

File: test-loop-tests/src/tests/resharding_cold_storage.rs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    let boundary_account: AccountId = "boundary".parse().unwrap();
+
+    let validators_spec = create_validators_spec(1, 0);
+    let mut clients = validators_spec_clients(&validators_spec);
+    let archival_id: AccountId = "archival".parse().unwrap();
+    clients.push(archival_id.clone());
+
+    let user_account: AccountId = "user0".parse().unwrap();
+
+    let genesis = TestLoopBuilder::new_genesis_builder()
```
Comment (staffik) [inline test-loop-tests/src/tests/resharding_cold_storage.rs:56]: Maybe use testloop Auto API, sth like this:
```
let base_epoch_config = TestEpochConfigBuilder::new()
      .shard_layout(base_shard_layout.clone())
      .epoch_length(EPOCH_LENGTH)
      .build();
  let (new_epoch_config, new_shard_layout) =
      derive_new_epoch_config_from_boundary(&base_epoch_config, &boundary_account);
  let epoch_config_store = EpochConfigStore::test(BTreeMap::from_iter([
      (PROTOCOL_VERSION - 1, Arc::new(base_epoch_config)),
      (PROTOCOL_VERSION, Arc::new(new_epoch_config)),
  ]));

  let mut env = TestLoopBuilder::new()
      .shard_layout(base_shard_layout.clone())
      .protocol_version(PROTOCOL_VERSION - 1)
      .epoch_length(EPOCH_LENGTH)
      .enable_archival_node(ArchivalKind::Cold)
      .epoch_config_store(epoch_config_store)
      .gc_num_epochs_to_keep(GC_NUM_EPOCHS_TO_KEEP)
      .build();
```

File: chain/chain/src/resharding/manager.rs [hunk 9]
```diff
@@ -246,6 +246,16 @@ impl ReshardingManager {
...
 
+            // Persist TrieChanges so that cold store can copy the resharding
+            // insertions. We store only insertions (no deletions) because
+            // deletions are not applied during resharding and would corrupt
+            // GC refcounts.
+            store_update.set_trie_changes(
+                *new_shard_uid,
+                block_hash,
+                &trie_changes.insertions_only(),
+            );
```
Comment (wacban) [inline chain/chain/src/resharding/manager.rs:257]: sanity check - the trie changes are stored with the resharding block (the last of old layout) but with the shard uids from after the resharding. It breaks an existing assumption but it makes sense. Sanity check is that this is what you intended and that you checked if the old assumption isn't relied on elsewhere.

File: chain/chain/src/garbage_collection.rs [hunk 10]
```diff
@@ -893,17 +893,36 @@ impl<'a> ChainStoreUpdate<'a> {
...
+        // shard UIDs NOT in this set come from resharding — they were persisted
+        // under child shard UIDs but the actual state insertions were applied
+        // under the parent shard UID.
+        let block_epoch_shard_uids: HashSet<ShardUId> = {
+            let block_header = self.get_block_header(&block_hash).expect("block header must exist");
+            let shard_layout = epoch_manager
+                .get_shard_layout(block_header.epoch_id())
+                .expect("epoch info must exist");
+            shard_layout.shard_uids().collect()
+        };
```
Comment (wacban) [inline chain/chain/src/garbage_collection.rs:908]: nit: I don't think it's necessary to wrap it in a block

File: chain/chain/src/garbage_collection.rs [hunk 11]
```diff
@@ -893,17 +893,36 @@ impl<'a> ChainStoreUpdate<'a> {
...
                 continue;
             };
             match gc_mode.clone() {
                 GCMode::Fork(tries) => {
-                    // If the block is on a fork, we delete the state that's the result of applying this block
-                    tries.revert_insertions(&trie_changes, shard_uid, store_update);
+                    // If the block is on a fork, we delete the state that's the result of applying this block.
+                    // Only revert TrieChanges from this block's own epoch (normal chunk processing).
+                    // Resharding entries (next-epoch shard UIDs) are skipped — see comment above.
+                    if block_epoch_shard_uids.contains(&shard_uid) {
```
Comment (wacban) [inline chain/chain/src/garbage_collection.rs:923]: Do we leak the TrieChanges in this case?

File: core/store/src/archive/cold_storage.rs [hunk 12]
```diff
@@ -56,13 +56,16 @@ pub trait ColdMigrationStore {
...
+/// the shard UID mapping is updated for cold store and the TrieChanges from
+/// the resharding block (previous block) are also copied.
 pub fn update_cold_db(
     cold_db: &ColdDB,
     hot_store: &Store,
     shard_layout: &ShardLayout,
     tracked_shards: &Vec<ShardUId>,
     height: &BlockHeight,
-    is_resharding_boundary: bool,
+    resharding_block_hash: Option<&[u8]>,
```
Comment (wacban) [inline core/store/src/archive/cold_storage.rs:68]: This is weird, why not CryptoHash?

File: core/store/src/archive/cold_storage.rs [hunk 13]
```diff
@@ -56,13 +56,16 @@ pub trait ColdMigrationStore {
 /// 1. add it to `DBCol::is_cold` list
 /// 2. define `DBCol::key_type` for it (if it isn't already defined)
 /// 3. add new clause in `get_keys_from_store` for new key types used for this column (if there are any)
+/// When `resharding_block_hash` is `Some`, it indicates a resharding boundary:
+/// the shard UID mapping is updated for cold store and the TrieChanges from
+/// the resharding block (previous block) are also copied.
```
Comment (wacban) [inline core/store/src/archive/cold_storage.rs:61]: sanity check: resharding block should be the last block of the old layout

File: core/store/src/archive/cold_storage.rs [hunk 14]
```diff
@@ -206,22 +211,26 @@ fn copy_state_from_store(
...
-            total_size += value.len();
-            tracing::trace!(target: "cold_store", pretty_key=?near_fmt::StorageKey(&key), "copying state node to colddb");
-            rc_aware_set(&mut transaction, DBCol::State, key, value);
+        // Look for TrieChanges at the current block hash (normal block processing)
+        // and, at resharding boundaries, also at the previous block hash (the
+        // resharding block where child shard state was created).
+        let block_hashes: [Option<&[u8]>; 2] = [Some(block_hash_key), resharding_block_hash];
+        for hash_key in block_hashes.into_iter().flatten() {
+            let key = join_two_keys(hash_key, &shard_uid_key);
+            let trie_changes: Option<TrieChanges> =
```
Comment (wacban) [inline core/store/src/archive/cold_storage.rs:222]: I don't think you need the type annotation (at least you shouldn't need both: here and in get_ser)

File: integration-tests/src/tests/client/cold_storage.rs [hunk 15]
```diff
@@ -170,15 +170,18 @@ fn test_storage_after_commit_of_cold_update() {
         let epoch_id = client.epoch_manager.get_epoch_id(block.hash()).unwrap();
         let shard_layout = client.epoch_manager.get_shard_layout(&epoch_id).unwrap();
         let shard_uids = shard_layout.shard_uids().collect();
-        let is_resharding_boundary =
-            client.epoch_manager.is_resharding_boundary(block.header().prev_hash()).unwrap();
+        let resharding_block_hash = client
+            .epoch_manager
+            .is_resharding_boundary(block.header().prev_hash())
+            .unwrap()
+            .then(|| block.header().prev_hash().as_bytes().to_vec());
```
Comment (wacban) [inline integration-tests/src/tests/client/cold_storage.rs:177]: nit: can you move prev_hash to a helper variable?

File: integration-tests/src/tests/client/cold_storage.rs [hunk 16]
```diff
@@ -331,15 +334,18 @@ fn test_cold_db_copy_with_height_skips() {
         let epoch_id = client.epoch_manager.get_epoch_id(&block_hash).unwrap();
         let shard_layout = client.epoch_manager.get_shard_layout(&epoch_id).unwrap();
         let shard_uids = shard_layout.shard_uids().collect();
-        let is_last_block_in_epoch =
-            client.epoch_manager.is_next_block_epoch_start(&block_hash).unwrap();
+        let resharding_block_hash = client
+            .epoch_manager
+            .is_resharding_boundary(block.header().prev_hash())
+            .unwrap()
+            .then(|| block.header().prev_hash().as_bytes().to_vec());
```
Comment (wacban) [inline integration-tests/src/tests/client/cold_storage.rs:341]: ditto

**Review by wacban — APPROVED** [review]
LGTM

File: chain/chain/src/garbage_collection.rs [hunk 17]
```diff
@@ -893,17 +893,36 @@ impl<'a> ChainStoreUpdate<'a> {
         store_update: &mut near_store::adapter::trie_store::TrieStoreUpdateAdapter<'_>,
     ) {
         let shard_uids_to_gc = self.get_shard_uids_to_gc(epoch_manager, &block_hash);
+
+        // Shard UIDs from the block's own epoch. TrieChanges for these shards
+        // come from normal block processing (chunk execution). TrieChanges for
+        // shard UIDs NOT in this set come from resharding — they were persisted
+        // under child shard UIDs but the actual state insertions were applied
+        // under the parent shard UID.
+        let block_epoch_shard_uids: HashSet<ShardUId> = {
```
Comment (staffik) [inline chain/chain/src/garbage_collection.rs:902]: Consider adding a debug_assert that child shard UIDs from the next epoch don't overlap with block_epoch_shard_uids. If that invariant ever breaks, revert_insertions would silently corrupt refcounts (decrementing under child UID while insertions went to parent UID).

**Review by staffik — APPROVED** [review]

File: chain/chain/src/garbage_collection.rs (hunk 11 above)
Comment (Wiezzel) [reply chain/chain/src/garbage_collection.rs:920]: No, we don't. TrieChanges row is always GC'ed. There was an issue with refcounts points out [here](https://github.com/near/nearcore/pull/15566#discussion_r3092159422), but it's fixed now.

File: chain/chain/src/garbage_collection.rs (hunk 7 above)
Comment (Wiezzel) [reply chain/chain/src/garbage_collection.rs]: Unnecessary, removed.

File: chain/chain/src/garbage_collection.rs (hunk 17 above)
Comment (Wiezzel) [reply chain/chain/src/garbage_collection.rs]: Added it, but it became messy and obscured the core logic, so I think we can live without this.
